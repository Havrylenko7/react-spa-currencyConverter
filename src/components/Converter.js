import React from 'react'; 
import { Form } from 'react-bootstrap';
import { Formik, Field } from 'formik';
import { useSelector } from 'react-redux';
import { BootCard, Option } from '../styled/Converter';

const Converter = () => {

  const rates = useSelector(state => state.myFirstReducer.rates);
  const ratesEntries = [...Object.entries(rates.eur)];

  return (
    <BootCard>
      <Formik
        initialValues={{ entries: ratesEntries, firstSelected: '', secondSelected: '' }}
        render={({ values }) => (
          <div>
            <Field as={Form.Select} size="lg" name="firstSelected">
              <Option>Pick currency</Option>
              {values.entries.map((item) => (
                <option key={item} value={`${item[1]}`}>
                  {item[0]}
                </option>
              ))}
            </Field>
            <Field as={Form.Select}  size="lg" name="secondSelected">
              <Option>Convert into</Option>
              {values.entries.map((item) => (
                <option key={item} value={`${item[1]}`}>
                  {item[0]}
                </option>
              ))}        
            </Field>
            <br />
            <div>
              {values.firstSelected.length && values.secondSelected.length > 0 ? (
                <div>{(values.secondSelected/values.firstSelected).toFixed(2)}</div>
              ) : null}
            </div>
          </div>
        )}
      />
    </BootCard>
  );
}

export default Converter;
