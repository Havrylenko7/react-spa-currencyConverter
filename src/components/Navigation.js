import { Nav } from 'react-bootstrap';
import { Link } from 'react-router-dom';

import { BootNav } from '../styled/Navigation';

const Navigation = () => 
  <Nav justify variant="pills">
    <Nav.Item>
      <BootNav as={Link} to="/rates">Exchange rates</BootNav>
    </Nav.Item>
    <Nav.Item>
      <BootNav as={Link} to="/converter">Currency converter</BootNav>
    </Nav.Item>
  </Nav>
;

export default Navigation;
