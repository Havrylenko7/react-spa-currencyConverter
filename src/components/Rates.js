import React from 'react';
import { ListGroup, Form} from 'react-bootstrap';
import { useSelector } from 'react-redux';
import { BootItem } from '../styled/Rates';

const Rates = () => {
  
  const [searchField, setSearchField] = React.useState('');
  const rates = useSelector(state => state.myFirstReducer.rates);
  const ratesEntries = [...Object.entries(rates.eur)];

  const handleChange = event => {
    setSearchField(event.target.value);
  };  

  return (
    <div>
      <Form.Control size="lg" type="search" placeholder="Search" onChange={handleChange}/>
      <br />
      <ListGroup>
        {ratesEntries
          .filter((item) => item[0].includes(searchField.toLowerCase()))
          .map((item) => 
            <BootItem key={item}>
              EUR to {item[0]} {item[1].toFixed(2)}
            </BootItem>
          )
        }
      </ListGroup>
    </div>
  );
}

export default Rates;
