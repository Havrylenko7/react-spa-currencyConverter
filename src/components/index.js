export { default as Converter } from './Converter';
export { default as Rates } from './Rates';
export { default as Navigation } from './Navigation';
