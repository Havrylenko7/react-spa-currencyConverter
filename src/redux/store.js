import { combineReducers, createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from '@redux-saga/core';

import { currencyReducer, saga } from './ducks/currency';

const sagaMiddleware = createSagaMiddleware();
const rootReducer = combineReducers({ currencyReducer });
const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(saga);

export default store;
