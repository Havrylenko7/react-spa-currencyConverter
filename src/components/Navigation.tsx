import { Nav } from 'react-bootstrap';

import { BootLink } from  '../styled/Navigation';

const Navigation: React.FC = () => (
  <Nav justify variant="pills">
    <Nav.Item>
      <Nav.Link as={BootLink} to="/rates" activeClassName="active">
        Rates
      </Nav.Link>
    </Nav.Item>
    <Nav.Item>
      <Nav.Link as={BootLink} to="/converter" activeClassName="active">
        Convert
      </Nav.Link>
    </Nav.Item>
  </Nav>
);

export default Navigation;
