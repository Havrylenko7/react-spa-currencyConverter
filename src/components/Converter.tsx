import React from 'react'; 
import { Formik, Field } from 'formik';

import { BootCard, CoverOption, BootField, BootForm, BootBody } from '../styled/Converter';
import { useTypedSelector } from '../redux/store';

const Converter: React.FC = () => {
  const rates: any = useTypedSelector(state => state.currencyReducer.rates);

  const ratesEntries = [...Object.entries(rates.eur)];

  ratesEntries.splice(18, 1);
  ratesEntries.splice(37, 1);

  return (
    <BootCard>
      <Formik
        initialValues={{ entries: ratesEntries, firstSelected: 0, secondSelected: 0, amount: 0}}
        onSubmit={() => {}}
        render={({ values }) => (
          <div>
            <Field as={BootField} name="firstSelected">
              <CoverOption>Pick currency</CoverOption>
              {values.entries.map((item: any) => (
                <option key={item} value={`${item[1]}`}>
                  {item[0]}
                </option>
              ))}
            </Field>
            <Field as={BootField} name="secondSelected">
              <CoverOption>Convert into</CoverOption>
              {values.entries.map((item: any) => (
                <option key={item} value={`${item[1]}`}>
                  {item[0]}
                </option>
              ))}        
            </Field>
            <Field as={BootForm} type="number" name="amount" min="1"/>
              {(values.firstSelected && values.secondSelected && values.amount) > 0 && (
                <BootBody>{(values.secondSelected/values.firstSelected*values.amount).toFixed(2)}</BootBody>
              )}
          </div>
        )}
      />
    </BootCard>
  );
};

export default Converter;
