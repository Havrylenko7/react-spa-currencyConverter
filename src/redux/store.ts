import { combineReducers, createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from '@redux-saga/core';
import { TypedUseSelectorHook, useSelector } from 'react-redux';

import { currencyReducer, saga } from './ducks/currency';

const sagaMiddleware = createSagaMiddleware();
export const rootReducer = combineReducers({ currencyReducer });
const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(saga);

export type RootState = ReturnType<typeof rootReducer>;
export const useTypedSelector: TypedUseSelectorHook<RootState>= useSelector;

export default store;
