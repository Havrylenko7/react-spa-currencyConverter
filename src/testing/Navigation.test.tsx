import { Provider } from 'react-redux';
import { shallow, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import { Navigation } from '../components';
import store from '../redux/store';

configure({ adapter: new Adapter() });

it('Navigation renders correctly', () => {
  const navigation = shallow(<Provider store={store}><Navigation /></Provider>)
  expect(navigation).toMatchSnapshot()
});
