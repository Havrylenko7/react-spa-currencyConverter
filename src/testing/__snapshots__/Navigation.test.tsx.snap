// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should match a snapshot 1`] = `ShallowWrapper {}`;
